src/async_impl/decoder.rs

```diff
@@ -43,29 +43,34 @@ pub(crate) struct Decoder {
     inner: Inner,
 }
 
+type PeekableIoStream = Peekable<IoStream>;
+
+#[cfg(any(feature = "gzip", feature = "brotli", feature = "deflate"))]
+type PeekableIoStreamReader = StreamReader<PeekableIoStream, Bytes>;
+
 enum Inner {
     /// A `PlainText` decoder just returns the response content as is.
     PlainText(super::body::ImplStream),
 
     /// A `Gzip` decoder will uncompress the gzipped response content before returning it.
     #[cfg(feature = "gzip")]
-    Gzip(FramedRead<GzipDecoder<StreamReader<Peekable<IoStream>, Bytes>>, BytesCodec>),
+    Gzip(Pin<Box<FramedRead<GzipDecoder<PeekableIoStreamReader>, BytesCodec>>>),
 
     /// A `Brotli` decoder will uncompress the brotlied response content before returning it.
     #[cfg(feature = "brotli")]
-    Brotli(FramedRead<BrotliDecoder<StreamReader<Peekable<IoStream>, Bytes>>, BytesCodec>),
+    Brotli(Pin<Box<FramedRead<BrotliDecoder<PeekableIoStreamReader>, BytesCodec>>>),
 
     /// A `Deflate` decoder will uncompress the deflated response content before returning it.
     #[cfg(feature = "deflate")]
-    Deflate(FramedRead<ZlibDecoder<StreamReader<Peekable<IoStream>, Bytes>>, BytesCodec>),
+    Deflate(Pin<Box<FramedRead<ZlibDecoder<PeekableIoStreamReader>, BytesCodec>>>),
 
     /// A decoder that doesn't have a value yet.
     #[cfg(any(feature = "brotli", feature = "gzip", feature = "deflate"))]
-    Pending(Pending),
+    Pending(Pin<Box<Pending>>),
```

Comment:
This would mean it always boxes, every time, right? Would it make sense to keep this box off? 80b does seem big, can that be shrunk?

Reply:
I am afraid there's no much I can do.

`Pending` contains `PeekableIoStream` and `DecoderType`, the former is 80B large and is alias to `Peekable<IoStream>`.

`IoStream` is just newtype wrapper of `body::ImplStream` which is 32B due to the `Bytes` stored inside.

[`Peekable`] stores `Fuse<IoStream>`, which is 40B due to [this](https://docs.rs/futures-util/latest/src/futures_util/stream/stream/fuse.rs.html#9-18) plus `Option<Result<Bytes, io::Error>>` which is 40B.

Unless we can remove that [`Peekable`], otherwise it would have to be this large.


[`Peekable`]: https://docs.rs/futures-util/latest/src/futures_util/stream/stream/peek.rs.html#14-27

Reply:
Oh, this is only if the headers indicate the body is encoded. So it's not _every_ body. This seems better. (I suppose technically we could share the box by restructuring the enum a bit, but meh.)

Reply:
> Oh, this is only if the headers indicate the body is encoded. So it's not every body. This seems better. (I suppose technically we could share the box by restructuring the enum a bit, but meh.)

One thing I don't understand is why do we need `Inner::Pending` in the first place.

In `impl Stream for Decoder`, branch for `Inner::Pending` calls `Pending::future` and uses the new `Inner` it returns to replace the current `self.inner`.

`Pending::future` simply waits for `PendingIoStream` to be ready without consuming anything, then take that `PendingIoStream` and constructs other decoders.

Why do we need this in the first place?
Why not construct the `Inner::Gzip` in `Decoder::gzip`?

Also I think `Peekable` can actually be removed since only `Pending` uses it.

Reply:
Well, I understand now why we need `Pending`: It's there to workaround `UnexpectedEof`.
If the stream is empty, then the decoder would return `UnexpectedEof`.